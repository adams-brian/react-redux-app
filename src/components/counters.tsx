import * as React from 'react';
import { Counter } from './counter';

interface ICountersProps {
  counters: number[];
  addCounter: () => void;
  increment: (index: number) => void;
  decrement: (index: number) => void;
  reset: (index: number) => void;
  remove: (index: number) => void;
}

/* tslint:disable:jsx-no-lambda */
export const Counters = (props: ICountersProps) => (
  <div className="counters-container">
    <h1>Counters</h1>
    <div className="counter-container">
      {props.counters.map((counter, index) => 
        (<Counter 
          key={index} /* not ideal, but it works in this case */
          counter={counter}
          increment={() => props.increment(index)}
          decrement={() => props.decrement(index)}
          reset={() => props.reset(index)}
          remove={() => props.remove(index)}
        />))}
      <div className="add-counter rounded" onClick={props.addCounter}>
        <span className="fa fa-plus" />
      </div>
    </div>
  </div>
);
