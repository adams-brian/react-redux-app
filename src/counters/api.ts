export const saveCounters = (counters: number[]) =>
  fetch('http://localhost:4000/counters', {
    body: JSON.stringify({counters}),
    headers: {
      'Accept': 'application/json, text/plain, */*',
      'Content-Type': 'application/json'
    },
    method: 'post',
  })

export const loadCounters = async (): Promise<number[]> => {
  const response = await fetch('http://localhost:4000/counters');
  const json = await response.json();
  return [...json.data];
}
