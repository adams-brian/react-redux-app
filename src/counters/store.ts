import { IAction } from '../store';

export interface IState {
  counters: number[];
}

/* tslint:disable:max-classes-per-file */
abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
export type CountersAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const actionCreators = {
  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),
};

function countersReducer(state: number[] = [], action: CountersAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

export const reducers = {
  counters: countersReducer
}
