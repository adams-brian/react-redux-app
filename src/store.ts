import { LocationChangeAction } from 'react-router-redux';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IAppState {
  loading: boolean;
  counters: number[];
  users: IUser[];
}

interface IAction {
  type: string;
}

/* tslint:disable:max-classes-per-file */
export const START_LOADING = '[App] START LOADING';
class StartLoading implements IAction {
  public readonly type = START_LOADING;
}
export const DONE_LOADING = '[App] DONE LOADING';
class DoneLoading implements IAction {
  public readonly type = DONE_LOADING;
}
type LoadingAction = StartLoading | DoneLoading;

abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
type CounterAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements IAction {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) {}
}
export const UPDATE_USER = '[Users] UPDATE';
class UpdateUser implements IAction {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) {}
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements IAction {
  public readonly type = DELETE_USER;
  constructor(public payload: string) {}
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements IAction {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) {}
}
type UserAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated;

export const actionCreators = {
  doneLoading: () => Object.assign({}, new DoneLoading()),
  startLoading: () => Object.assign({}, new StartLoading()),

  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),

  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

function loadingReducer(state: boolean = false, action: LoadingAction) {
  switch (action.type) {
    case START_LOADING:
      return true;
    case DONE_LOADING:
      return false;
    default:
      return state;
  }
}

function countersReducer(state: number[] = [], action: CounterAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

const usersReducer = (state: IUser[] = [], action: UserAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const reducers = {
  counters: countersReducer,
  loading: loadingReducer,
  users: usersReducer
};

export type AppAction = LoadingAction | CounterAction | UserAction | LocationChangeAction;
