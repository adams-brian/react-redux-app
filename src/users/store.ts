import { Action, combineReducers } from 'redux';

export interface IUser {
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IState {
  users: {
    error: string;
    list: IUser[];
    loaded: boolean;
  }
}

/* tslint:disable:max-classes-per-file */
export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements Action {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) { }
}
export const UPDATE_USER = '[Users] UPDATE';
export class UpdateUser implements Action {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) { }
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements Action {
  public readonly type = DELETE_USER;
  constructor(public payload: string) { }
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements Action {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) { }
}
export const LOAD_USERS = '[Users] LOAD USERS';
export class LoadUsers implements Action {
  public readonly type = LOAD_USERS;
}
export const USERS_ERROR = '[Users] USERS ERROR';
export class UsersError implements Action {
  public readonly type = USERS_ERROR;
  constructor(public payload: Error) { }
}
export const USERS_LOADED = '[Users] USERS LOADED';
export class UsersLoaded implements Action {
  public readonly type = USERS_LOADED;
}
export type UsersAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated | LoadUsers | UsersError | UsersLoaded;

export const actionCreators = {
  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  loadUsers: () => Object.assign({}, new LoadUsers()),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersError: (err: Error) => Object.assign({}, new UsersError(err)),
  usersLoaded: () => Object.assign({}, new UsersLoaded()),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

export const listReducer = (state: IUser[] = [], action: UsersAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const loadedReducer = (state: boolean = false, action: UsersAction) => {
  switch (action.type) {
    case USERS_LOADED:
      return true;
    default:
      return state;
  }
}

export const errorReducer = (state: string = '', action: UsersAction) => {
  switch (action.type) {
    case USERS_ERROR:
      return action.payload.message;
    default:
      return state;
  }
}

export const reducers = {
  users: combineReducers({
    error: errorReducer,
    list: listReducer,
    loaded: loadedReducer
  })
};
