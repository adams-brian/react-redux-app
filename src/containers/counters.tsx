import { connect } from 'react-redux';
import { AppState, actionCreators } from '../store';

import { Counters as UnwrappedCounters } from '../components/counters';

export const Counters = connect(
  (state: AppState) => 
    ({ counters: state.counters }),
  { 
    addCounter: actionCreators.addCounter,
    increment: actionCreators.increment,
    decrement: actionCreators.decrement,
    reset: actionCreators.reset,
    remove: actionCreators.removeCounter
  }
  )(UnwrappedCounters);
  