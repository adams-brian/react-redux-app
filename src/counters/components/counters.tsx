import * as React from 'react';
import { connect } from 'react-redux';

import { actionCreators, IAppState } from '../../store';
import Counter from './counter';

interface ICountersProps {
  counters: number[];
  addCounter: () => void;
  increment: (index: number) => void;
  decrement: (index: number) => void;
  reset: (index: number) => void;
  remove: (index: number) => void;
}

/* tslint:disable:jsx-no-lambda */
export const Counters = (props: ICountersProps) => (
  <div className="counters-container">
    <h1>Counters</h1>
    <div className="counter-container">
      {props.counters.map((counter, index) => 
        (<Counter 
          key={index} /* not ideal, but it works in this case */
          counter={counter}
          increment={() => props.increment(index)}
          decrement={() => props.decrement(index)}
          reset={() => props.reset(index)}
          remove={() => props.remove(index)}
        />))}
      <div className="add-counter rounded" onClick={props.addCounter}>
        <span className="fa fa-plus" />
      </div>
    </div>
  </div>
);

export default connect(
  (state: IAppState) => 
    ({ counters: state.counters }),
  { 
    addCounter: actionCreators.addCounter,
    decrement: actionCreators.decrement,
    increment: actionCreators.increment,
    remove: actionCreators.removeCounter,
    reset: actionCreators.reset
  }
)(Counters);
  
