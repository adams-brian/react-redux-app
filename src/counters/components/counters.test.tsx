import { shallow } from 'enzyme';
import * as React from 'react';
import * as sinon from 'sinon';

import { actionCreators } from '../store';
import { Counters, mapDispatchToProps, mapStateToProps } from './counters';

describe("Counters", () => {

  const list = [5,1,4,2,3];
  const loaded = true;
  const error = '';
  const counters = { list, loaded, error };

  describe("component", () => {

    type propFunction = (index: number) => void;
    let element: JSX.Element;
    let addCounter: sinon.SinonSpy;
    let increment: sinon.SinonSpy;
    let decrement: sinon.SinonSpy;
    let reset: sinon.SinonSpy;
    let remove: sinon.SinonSpy;
    
    beforeEach(() => {
      addCounter = sinon.spy();
      increment = sinon.spy();
      decrement = sinon.spy();
      reset = sinon.spy();
      remove = sinon.spy();

      element = <Counters
        counters={[...list]}
        addCounter={addCounter}
        increment={increment}
        decrement={decrement}
        reset={reset}
        remove={remove}
      />
    });

    it('renders as expected', () => {
      const component = shallow(element);
      expect(component).toMatchSnapshot();
    });

    it('calls addCounter', () => {
      const component = shallow(element);
      expect(addCounter.called).toBe(false);
      component.find('.add-counter').simulate('click');
      expect(addCounter.calledOnce).toBe(true);
    });

    it('calls increment', () => {
      const component = shallow(element);
      expect(increment.called).toBe(false);
      const index = 1;
      (component.find('Counter').at(index).prop('increment') as propFunction)(index);
      expect(increment.calledOnceWithExactly(index)).toBe(true);
    });

    it('calls decrement', () => {
      const component = shallow(element);
      expect(decrement.called).toBe(false);
      const index = 2;
      (component.find('Counter').at(index).prop('decrement') as propFunction)(index);
      expect(decrement.calledOnceWithExactly(index)).toBe(true);
    });

    it('calls reset', () => {
      const component = shallow(element);
      expect(reset.called).toBe(false);
      const index = 3;
      (component.find('Counter').at(index).prop('reset') as propFunction)(index);
      expect(reset.calledOnceWithExactly(index)).toBe(true);
    });

    it('calls remove', () => {
      const component = shallow(element);
      expect(remove.called).toBe(false);
      const index = 4;
      (component.find('Counter').at(index).prop('remove') as propFunction)(index);
      expect(remove.calledOnceWithExactly(index)).toBe(true);
    });

  });

  describe('connection', () => {

    it('maps state to props', () => {
      expect(mapStateToProps({ counters })).toEqual({ counters: list, loaded, error });
    });
  
    it('maps dispatch to props', () => {
      expect(mapDispatchToProps).toEqual({ 
        addCounter: actionCreators.addCounter,
        decrement: actionCreators.decrement,
        increment: actionCreators.increment,
        load: actionCreators.loadCounters,
        remove: actionCreators.removeCounter,
        reset: actionCreators.reset
      });
    });

  });

});
