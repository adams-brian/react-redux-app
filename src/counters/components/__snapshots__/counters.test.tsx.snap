// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counters component renders as expected 1`] = `
<div
  className="counters-container"
>
  <h1>
    Counters
  </h1>
  <h3>
    Count: 
    5
  </h3>
  <div
    className="counter-container"
  >
    <Counter
      counter={5}
      decrement={[Function]}
      increment={[Function]}
      index={0}
      key="0"
      remove={[Function]}
      reset={[Function]}
    />
    <Counter
      counter={1}
      decrement={[Function]}
      increment={[Function]}
      index={1}
      key="1"
      remove={[Function]}
      reset={[Function]}
    />
    <Counter
      counter={4}
      decrement={[Function]}
      increment={[Function]}
      index={2}
      key="2"
      remove={[Function]}
      reset={[Function]}
    />
    <Counter
      counter={2}
      decrement={[Function]}
      increment={[Function]}
      index={3}
      key="3"
      remove={[Function]}
      reset={[Function]}
    />
    <Counter
      counter={3}
      decrement={[Function]}
      increment={[Function]}
      index={4}
      key="4"
      remove={[Function]}
      reset={[Function]}
    />
    <div
      className="add-counter rounded"
      onClick={[Function]}
    >
      <span
        className="fa fa-plus"
      />
    </div>
  </div>
</div>
`;
